internal/streams/streams.go

```diff
@@ -41,22 +42,62 @@ func NewLimited(maxStreams int) *IDGenerator {
 		NumStreams: maxStreams,
 		streams:    streams,
 		numBuckets: uint32(buckets),
+		offsetStep: uint32(findOffsetStep(buckets)),
 		offset:     uint32(buckets) - 1,
 	}
 }
 
+// findOffsetStep returns the offset step to use for the given number of buckets.
+// `offset` provides a way to reduce collision by offsetting the starting point
+// `offsetStep` should be small enough to provide some cycles before reaching `buckets`,
+// and big enough to have significant gap between invocations to ensure that collision chance is actually reduced.
+// and it should not divide `buckets` without remainder, to ensure it does not land on the same indexes all the time.
+func findOffsetStep(buckets int) int {
+	switch buckets {
+	case 1:
+		return 1
+	case 2:
+		return 1
+	case 3:
+		return 2
+	}
+	if buckets < 8 {
+		offsetStep := buckets / 2
+		if offsetStep == 0 {
+			return 1
+		}
+		if buckets%2 != 0 {
+			return offsetStep
+		}
+		return offsetStep + 1
+	}
+	for _, offsetStep := range []int{13, 11, 9, 7, 5, 3} {
+		if buckets/offsetStep < 4 {
+			// Ensure that it provides at least 4 cycles before reaching `buckets`
+			continue
+		}
+		if buckets%offsetStep != 0 {
+			return offsetStep
+		}
+
+		offsetStep++
+		if buckets%offsetStep != 0 {
+			return offsetStep
+		}
+	}
+	if buckets%2 != 0 {
+		return 2
+	}
+	return 1
+}
```

Comment:
This code has a lot of non-obvious conditionals and magic numbers. There should by comments explaining not only what the code does, but also how and why.

Reply:
Agree, it suppose to reduce collissions in the case when buckets are highly populated, but due to it's complexiity I start thinking that it does not worth it, let's just remove it and go back to +1